src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp

```diff
@@ -0,0 +1,244 @@
+    std::vector<double> qi(beta.size(),
+                           0.0); // qi is the loss for each class like in output
+                                 // we have [0,1,2] so loss for each class
+    for (int j = 0; j < beta.size(); j++)
+    { // each class iteration
+      for (int k = 0; k < xiHat.size(); k++)
+      { // in one class each weight iteration
+        // qi =   [ __ , __ , __ , __....]
+        //          [ __ , __ , __ , __....]  * [ __,__ ,__ ,.....]
+        //          [ __ , __ , __ , __....]        xiHat
+        //          [ __ , __ , __ , __....]
+        //              beta
+        qi[j] += beta[j][k] * xiHat[k];
+      }
+    }
+    qi = softmax(qi);             // qi = z[i] = x[i]*beta[i];
+    loss += crossEntropy(yi, qi); // Computing loss using z[i](computed values)
+                                  // and y[i](actual values)
+  }
+  return loss; // finally returning loss
+};
+
+template<class T>
+// This function performs logistic regression stochastic gradient descent
+std::vector<std::vector<double>>
+LogisticRegression<T>::logRegSgd(std::vector<std::vector<double>> x,
+                                 std::vector<std::vector<int>> y, double alpha,
+                                 int numEpochs, bool verbose, int batchSize)
+{                               // PERFORMS MULTICLASS LOGISTIC REGRESSION
+  int n = x.size();             // Rows in training vector
+  int d = x[0].size();          // Number of features
+  int numClasses = y[0].size(); // Output classes
+  std::vector<std::vector<double>> beta(
+      numClasses,
+      std::vector<double>(
+          d, 0.0)); // Initializing weights (output_classes * features)
+  std::vector<double> lVals(numEpochs);
+  for (int i = 0; i < numEpochs;
+       i++) // loop running for number of epochs specified
+  {
+    double l = EvalL(x, y, beta); // computes the loss when beta is
+    // [__,__,......]
+    // [__,__,......]
+    // [__,__,......]
+    lVals.push_back(l);
+
+    if (verbose) // decides whether to print epoch number and current loss
+                 // computed by EvalL
+    {
+      std::cout << "Epoch: " << i << " "
+                << "Loss: " << l << std::endl;
+    }
+
+    std::vector<int> prm(n); // Making vector of size 'n' so that we can
+                             // randomly shuffle number from 0-(n-1) and choose
+                             // values in random order to perform regression
+    for (int j = 0; j < n; j++)
+    {
+      prm[j] = j;
+    }
+
+    std::random_shuffle(prm.begin(),
+                        prm.end()); // random shuffling of prm vector
+
+    int countCompleted = 0;
+
+    std::vector<std::vector<double>> gradLi( // Initializing gradient descent
+        numClasses, std::vector<double>(x[0].size(), 0.0));
+
+    while (!prm.empty()) // till prm not becomes empty keep iterating over the
+                         // while loop
+    {
+      if ((prm.size() + countCompleted) <
```

Comment:
This isn't how it actually works.
In each epoch, you need to choose a random value (the starting point of that batch) and take a batch of the given size from that....many training inputs might be considered multiple times in such a case.